examples/raja-teams.cpp

```diff
@@ -0,0 +1,366 @@
+  const int N = NThreads * NBlocks;
+#else
+  const int NThreads = 1;
+  const int N = NThreads * NBlocks;
+#endif
+
+  //
+  // Allocate and initialize matrix data.
+  //
+  double *A = memoryManager::allocate<double>(N * N);
+  double *B = memoryManager::allocate<double>(N * N);
+  double *C = memoryManager::allocate<double>(N * N);
+
+  for (int row = 0; row < N; ++row) {
+    for (int col = 0; col < N; ++col) {
+      A(row, col) = row;
+      B(row, col) = col;
+    }
+  }
+
+  std::cout << "\n Running RAJA-Teams V2-version of matrix multiplication...\n";
+  std::cout << "  N = " << N << std::endl;
+
+  for (int exec_place = 0; exec_place < (int)RAJA::NUM_PLACES; ++exec_place) {
+    RAJA::ExecPlace select_cpu_or_gpu = (RAJA::ExecPlace)exec_place;
+    // auto select_cpu_or_gpu = RAJA::HOST;
+    // auto select_cpu_or_gpu = RAJA::DEVICE;
+
+    /*
+     * launch just starts a "kernel" it's doesn't provide any looping.
+     *
+     * The first argument determines which policy should be executed,
+     *
+     * The second argument is the number of teams+threads needed for each of the
+     * policies.
+     *
+     * Third argument is the lambda for the policy.
+     *
+     *
+     * The lambda takes a "resource" object, which has the teams+threads and
+     * policy selection information.
+     */
+
+    //========================
+    // Upper triangular pattern
+    //========================
+    const int N_tri = 5;
+    RAJA::launch<launch_policy>(
+        select_cpu_or_gpu,
+        RAJA::ResourceList{
+            RAJA::Resources(RAJA::Threads(N_tri)),
+            RAJA::Resources(RAJA::Teams(N_tri), RAJA::Threads(N_tri))},
+        [=] RAJA_HOST_DEVICE(RAJA::LaunchContext ctx) {
+          RAJA::loop<teams0>(ctx, RAJA::RangeSegment(0, N_tri), [=](int i) {
+            // do a matrix triangular pattern
+            RAJA::loop<threads0>(ctx, RAJA::RangeSegment(i, N_tri), [=](int j) {
+              printf("i=%d, j=%d\n", i, j);
+            });  // loop j
+          });    // loop i
+        });      // kernel
+
+    //========================
+    // Matrix-Matrix Multiplication Example
+    //========================
+    // Set up Teams/Threads
+
+    RAJA::RangeSegment TeamRange(0, NBlocks);
+
+    RAJA::launch<launch_policy>(select_cpu_or_gpu,
+          RAJA::ResourceList{
+          RAJA::Resources(RAJA::Teams(NBlocks, NBlocks), RAJA::Threads(NThreads, NThreads))},
+        [=] RAJA_HOST_DEVICE(RAJA::LaunchContext ctx) {
+          //
+          // Loop over teams
+          //
+          //RAJA::loop<teams1>(ctx, RAJA::RangeSegment(0, NBlocks), [&](int by) {
+          //RAJA::loop<teams0>(ctx, RAJA::RangeSegment(0, NBlocks), [&](int bx) {
+
+          RAJA::loop<teams01>(ctx, TeamRange, TeamRange, [&](int bx, int by) {
+
+                  TEAM_SHARED double As[NThreads][NThreads];
+                  TEAM_SHARED double Bs[NThreads][NThreads];
+                  TEAM_SHARED double Cs[NThreads][NThreads];
```

Comment:
Something else that remains is how to replace using a macro here... We could replace with some sort of shared memory object? Thoughts @camierjs ?